common/sdk/factory.go

```diff
@@ -56,27 +61,45 @@ func NewClientFactory(hostPort string, tlsConfig *tls.Config, metricsHandler *Me
 		hostPort:       hostPort,
 		tlsConfig:      tlsConfig,
 		metricsHandler: metricsHandler,
+		once:           &sync.Once{},
 	}
 }
 
-func (f *clientFactory) NewClient(namespaceName string, logger log.Logger) (sdkclient.Client, error) {
-	sdkClient, err := sdkclient.NewClient(sdkclient.Options{
-		HostPort:       f.hostPort,
-		Namespace:      namespaceName,
-		MetricsHandler: f.metricsHandler,
-		Logger:         log.NewSdkLogger(logger),
-		ConnectionOptions: sdkclient.ConnectionOptions{
-			TLS:                f.tlsConfig,
-			DisableHealthCheck: true,
-		},
-	})
+func (f *clientFactory) NewClient(namespaceName string, logger log.Logger) sdkclient.Client {
+	var client sdkclient.Client
+
+	// Retry for up to 1m, handles frontend service not ready
+	err := backoff.Retry(func() error {
+		sdkClient, err := sdkclient.NewClient(sdkclient.Options{
+			HostPort:       f.hostPort,
+			Namespace:      namespaceName,
+			MetricsHandler: f.metricsHandler,
+			Logger:         log.NewSdkLogger(logger),
+			ConnectionOptions: sdkclient.ConnectionOptions{
+				TLS: f.tlsConfig,
+			},
+		})
+		if err != nil {
+			return fmt.Errorf("unable to create SDK client: %w", err)
+		}
+
+		client = sdkClient
+		return nil
+	}, common.CreateSdkClientFactoryRetryPolicy(), common.IsContextDeadlineExceededErr)
+
 	if err != nil {
-		return nil, fmt.Errorf("unable to create SDK client: %w", err)
+		logger.Fatal(
+			"error getting system sdk client",
+			tag.Error(err),
+		)
 	}
 
-	return sdkClient, nil
+	return client
```

Comment:
I think we should return error from NewClient(), the fatal should only happen in GetSystemClient(). The reason is the NewClient() could be called in other place like the batcher that you pointed out, and the namespace there is from user input which may not exists and we should not crash process because of that.

Reply:
That seems like a missed input validation in tctl before starting the workflow that executes this activity.

Agree though if that condition exists today, a crash is not the correct outcome.